File: Solutions/Product of Array Except Self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> ans(n, 1);
        
        // find cumulative left product
        int prod = 1;
        for (int i=0; i<n-1; i++) {
            prod *= nums[i];
            ans[i+1] = prod;
        }
        
        // find cumulative right product
        prod = 1;
        for (int i=n-1; i>0; i--) {
            prod *= nums[i];
            ans[i-1] *= prod;
        }
        
        return ans;
    }
};

/*
ans[i] = cumulative product from the left (0 to i-1) * 
         cumulative product from the right (n-1 to i+1)

space complexity optimization:
    - we first accumulate the left cumulative products in ans.
    - we then multiply these by the right cumulative products.
    - hence, we do not use any additional space to store the products.
    
time complexity: O(n)
space complexity: O(1)

https://leetcode.com/problems/product-of-array-except-self/
*/

File: Blind75/medium/Product of Array Except Self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        out = [1 for _ in range(n)]
        
        # left cumulative product
        product = 1
        for i in range(0, n-1):
            product *= nums[i]
            out[i+1] *= product
        
        # right cumulative product
        product = 1
        for i in range(n-1, 0, -1):
            product *= nums[i]
            out[i-1] *= product
        
        return out

"""
product except self at an index i = 
    left cumulative product(0, i-1) * 
    right cumulative product(i+1, n-1)

- calculate left and right cumulative products separately
- multiply them with the output array directly to use O(1)
  space.

time complexity: O(n)
space complexity: O(1) excluding the output
"""